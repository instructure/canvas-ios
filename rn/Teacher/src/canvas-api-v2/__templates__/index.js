// @flow
export * from './submission'
export * from './user'
export * from './assignment'
export * from './course'
export * from './group'
export * from './section'
export * from './groupSet'

// QUERIES
export * from '../queries/__templates__/SubmissionList'
export * from '../queries/__templates__/ContextCard'
