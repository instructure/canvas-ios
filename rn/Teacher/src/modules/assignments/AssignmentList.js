//
// Copyright (C) 2017-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

/**
* Launching pad for navigation for a single course
* @flow
*/

import React, { Component } from 'react'
import { connect } from 'react-redux'
import {
  View,
  StyleSheet,
  ActionSheetIOS,
  SectionList,
  NativeModules,
  ActivityIndicator,
} from 'react-native'
import i18n from 'format-message'

import AssignmentListActions from './actions'
import CourseActions from '../courses/actions'
import { mapStateToProps, type AssignmentListProps } from './map-state-to-props'
import refresh from '../../utils/refresh'

import { isTeacher } from '../app'
import SectionHeader from '../../common/components/rows/SectionHeader'
import AssignmentListRowView from './components/AssignmentListRow'
import { LinkButton } from '../../common/buttons'
import { Heading1, Heading2, Text } from '../../common/text'
import Screen from '../../routing/Screen'
import { type TraitCollection } from '../../routing/Navigator'
import { isRegularDisplayMode } from '../../routing/utils'
import ActivityIndicatorView from '../../common/components/ActivityIndicatorView'
import ListEmptyComponent from '../../common/components/ListEmptyComponent'
import { getGradesForGradingPeriod } from '../../canvas-api'
import { logEvent } from '../../common/CanvasAnalytics'
import { featureFlagEnabled } from '../../common/feature-flags'

type State = {
  currentFilter: {
    index?: number,
    title: string,
  },
  filterApplied: boolean,
  currentScore: ?number,
  loadingGrade: boolean,
  gradeError: boolean,
  selectedRowID: ?string,
}

const { NativeAccessibility, SiriShortcutManager } = NativeModules

export class AssignmentList extends Component<AssignmentListProps, State> {
  state: State
  isRegularScreenDisplayMode: boolean
  data: any = []
  didSelectFirstItem = false

  static defaultProps = {
    ListRow: AssignmentListRowView,
  }

  state = {
    currentFilter: { title: i18n('All Grading Periods') },
    filterApplied: false,
    currentScore: this.props.currentScore,
    loadingGrade: false,
    gradeError: false,
    selectedRowID: null,
  }

  componentDidMount () {
    this.showCurrentPeriod()
  }

  componentDidUpdate (prevProps: AssignmentListProps) {
    if (
      prevProps.gradingPeriods.length !== this.props.gradingPeriods.length ||
      prevProps.currentGradingPeriodID !== this.props.currentGradingPeriodID
    ) {
      this.showCurrentPeriod() // can only updateFilter properly after props is updated
    }
  }

  showCurrentPeriod () {
    const index = this.props.gradingPeriods.findIndex(({ id }) =>
      id === this.props.currentGradingPeriodID
    )
    if (index >= 0) this.updateFilter(index)
  }

  componentWillMount () {
    this.onTraitCollectionChange()
  }

  componentWillReceiveProps (nextProps: AssignmentListProps) {
    if (nextProps.assignmentGroups.length && nextProps.gradingPeriods.length) {
      NativeAccessibility.refresh()
    }

    if (nextProps.currentScore !== this.props.currentScore) {
      this.setState({ currentScore: nextProps.currentScore })
    }
  }

  componentWillUnmount () {
    const currentScore = this.state.currentScore || 0
    if (this.props.showTotalScore && currentScore > 90) {
      NativeModules.AppStoreReview.handleSuccessfulSubmit()
    }
  }

  onTraitCollectionChange () {
    this.props.navigator.traitCollection((traits) => { this.traitCollectionDidChange(traits) })
  }

  traitCollectionDidChange (traits: TraitCollection) {
    this.isRegularScreenDisplayMode = isRegularDisplayMode(traits)
    if (!this.isRegularScreenDisplayMode) {
      this.didSelectFirstItem = false
    }
    this.selectFirstListItemIfNecessary()
  }

  selectFirstListItemIfNecessary () {
    let assignment = null
    const isModal = this.props.navigator.isModal
    if (!this.props.doNotSelectFirstItem && !this.didSelectFirstItem && this.isRegularScreenDisplayMode && !isModal && (assignment = this.firstAssignmentInList())) {
      this.selectedAssignment(assignment)
      this.didSelectFirstItem = true
    }
  }

  firstAssignmentInList (): ?Assignment {
    if (this.data.length > 0 && this.data[0].assignments.length > 0) {
      return this.data[0].assignments.sort((a, b) => a.position - b.position)[0]
    }
    return null
  }

  prepareListData () {
    return this.props.assignmentGroups.map(group => {
      let gradingPeriodFilter
      if (this.state.currentFilter.index != null) {
        gradingPeriodFilter = this.props.gradingPeriods[this.state.currentFilter.index]
      }
      let assignments = this.state.filterApplied
        ? group.assignments.filter(({ id }) => gradingPeriodFilter.assignmentRefs.includes(id))
        : group.assignments

      if (assignments.length) {
        return {
          key: group.id,
          ...group,
          data: assignments.slice().sort((a, b) => a.position - b.position),
        }
      }
    }).filter(item => item)
  }

  renderRow = ({ item, index }: { item: Assignment, index: number }) => {
    let selected = this.isRegularScreenDisplayMode && this.state.selectedRowID === item.id
    let ListRow = this.props.ListRow
    return (
      // $FlowFixMe - for some reason flow doesn't like the default prop
      <ListRow
        assignment={item}
        tintColor={this.props.courseColor}
        onPress={this.selectedAssignment}
        key={index}
        selected={selected}
        user={this.props.user}
      />
    )
  }

  renderSectionHeader = ({ section }: any) => {
    return <SectionHeader title={section.name} key={section.key} top={section.position === 1} />
  }

  selectedAssignment = (assignment: Assignment) => {
    logEvent('assignment_selected')
    this.props.updateCourseDetailsSelectedTabSelectedRow(assignment.id)
    this.setState({ selectedRowID: assignment.id })
    if (assignment.quiz_id && !featureFlagEnabled('newStudentAssignmentView')) {
      this.props.navigator.show(`/courses/${assignment.course_id}/quizzes/${assignment.quiz_id}`)
    } else if (assignment.discussion_topic && isTeacher()) {
      this.props.navigator.show(`/courses/${assignment.course_id}/discussion_topics/${assignment.discussion_topic.id}`)
    } else {
      this.props.navigator.show(assignment.html_url)
    }
  }

  updateGradeForGradingPeriod = async (gradingPeriod: GradingPeriod) => {
    try {
      this.setState({ loadingGrade: true, gradeError: false })
      let grades = await getGradesForGradingPeriod(this.props.courseID, 'self', gradingPeriod.id)
      this.setState({ loadingGrade: false, currentScore: grades.current_score })
    } catch (err) {
      console.error('Error loading grade', err)
      this.setState({ loadingGrade: false, gradeError: true })
    }
  }

  clearFilter = () => {
    this.setState({
      currentFilter: { title: i18n('All Grading Periods') },
      filterApplied: false,
      currentScore: this.props.currentScore,
    })
  }

  applyFilter = () => {
    let buttons = this.props.gradingPeriods.map(({ title }) => title).concat(i18n('Cancel'))
    ActionSheetIOS.showActionSheetWithOptions({
      options: buttons,
      cancelButtonIndex: buttons.length - 1,
      title: i18n('Filter by:'),
    }, this.updateFilter)
  }

  updateFilter = (index: number) => {
    // don't do anything if the user hits cancel
    if (index === this.props.gradingPeriods.length) return

    // always get assignment info for grading period, since it might be shared
    this.props.refreshAssignmentList(this.props.courseID, this.props.gradingPeriods[index].id, true)

    // get the grade for the current grading period
    if (this.props.showTotalScore) {
      this.updateGradeForGradingPeriod(this.props.gradingPeriods[index])
    }

    this.setState({
      currentFilter: {
        title: this.props.gradingPeriods[index].title,
        index,
      },
      filterApplied: true,
    })
  }

  toggleFilter = () => {
    if (this.state.filterApplied) {
      this.clearFilter()
    } else {
      this.applyFilter()
    }
  }

  render () {
    if (this.props.pending && !this.props.refreshing) {
      return <ActivityIndicatorView />
    }
    if (this.data.length === 0) {
      this.data = this.prepareListData()
      this.selectFirstListItemIfNecessary()
    } else {
      this.data = this.prepareListData()
    }

    if (this.props.showGrades) {
      SiriShortcutManager.donateSiriShortcut({ 'identifier': 'com.instructure.siri.shortcut.getgrades', 'url': `/courses/${this.props.courseID}/grades`, 'name': this.props.courseCode })
    }

    return (
      <Screen
        title={this.props.screenTitle}
        onTraitCollectionChange={this.onTraitCollectionChange.bind(this)}
        subtitle={this.props.courseName}
        navBarColor={this.props.courseColor}
        navBarStyle='dark'
        testID='assignment-list'
      >
        <View style={styles.container} testID='assignment-list.container-view'>
          <View style={styles.header}>
            <View style={styles.gradingPeriodHeader} testID='assignment-list.filter-header-view'>
              <Heading1 style={styles.headerTitle} testID='assignment-list.filter-title-lbl'>{this.state.currentFilter.title}</Heading1>
              {this.props.gradingPeriods.length > 0 &&
                <LinkButton testID='assignment-list.filter' onPress={this.toggleFilter} style={styles.filterButton}>
                  {this.state.filterApplied
                    ? i18n('Clear filter')
                    : i18n('Filter')}
                </LinkButton>
              }
            </View>
            {this.props.showTotalScore &&
              <View style={styles.gradeHeader} testID='assignment-list.total-grade'>
                <Heading2>{i18n('Total Grade:')}</Heading2>
                {this.state.loadingGrade
                  ? <ActivityIndicator />
                  : this.state.currentScore
                    ? <Text>{i18n.number(this.state.currentScore / 100, 'percent')}</Text>
                    : <Text>{i18n('N/A')}</Text>
                }
              </View>
            }
          </View>
          <SectionList
            testID='assignment-list.list'
            sections={this.data}
            renderItem={this.renderRow}
            renderSectionHeader={this.renderSectionHeader}
            refreshing={this.props.refreshing}
            onRefresh={this.refresh}
            keyExtractor={(item, index) => item.id}
            ListEmptyComponent={
              <ListEmptyComponent title={i18n('There are no assignments to display.')} />
            }
          />
        </View>
      </Screen>
    )
  }

  refresh = () => {
    if (this.state.currentFilter.index != null) {
      // Refresh total for selected grading period
      const gradingPeriod = this.props.gradingPeriods[this.state.currentFilter.index]
      this.updateGradeForGradingPeriod(gradingPeriod)
    }

    this.props.refresh()
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  header: {
    borderBottomWidth: StyleSheet.hairlineWidth,
    borderBottomColor: 'lightgrey',
    paddingTop: global.style.defaultPadding,
    paddingBottom: global.style.defaultPadding / 2,
    paddingHorizontal: global.style.defaultPadding,
  },
  gradingPeriodHeader: {
    flexDirection: 'row',
    alignItems: 'flex-end',
    justifyContent: 'space-between',
  },
  headerTitle: {
    flex: 1,
  },
  filterButton: {
    marginBottom: 1,
  },
  gradeHeader: {
    flexDirection: 'row',
    alignItems: 'flex-end',
    justifyContent: 'space-between',
  },
})

const Refreshed = refresh(
  props => {
    props.refreshAssignmentList(props.courseID)
    props.refreshGradingPeriods(props.courseID)
  },
  props => props.assignmentGroups.length === 0 || props.gradingPeriods.length === 0,
  props => Boolean(props.pending),
)(AssignmentList)
const Connected = connect(mapStateToProps, { ...AssignmentListActions, ...CourseActions })(Refreshed)
export default (Connected: Component<AssignmentListProps, State>)
