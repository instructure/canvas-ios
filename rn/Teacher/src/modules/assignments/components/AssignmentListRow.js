//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/**
* @flow
*/

import React, { PureComponent } from 'react'
import {
  View,
  StyleSheet,
} from 'react-native'

import i18n from 'format-message'
import { formattedDueDateWithStatus } from '../../../common/formatters'
import AccessIcon from '../../../common/components/AccessIcon'
import AssignmentDates from '../../../common/AssignmentDates'
import { Text } from '../../../common/text'
import Row from '../../../common/components/rows/Row'
import Images from '../../../images'

type Props = {
  assignment: Assignment,
  tintColor: string,
  onPress: (Assignment) => void,
  selected: boolean,
}

export default class AssignmentListRow extends PureComponent<Props> {
  onPress = () => {
    const assignment = this.props.assignment
    this.props.onPress(assignment)
  }

  dueDate (assignment: Assignment): string {
    const dates = new AssignmentDates(assignment)

    if (dates.availabilityClosed()) {
      return i18n('Availability: Closed')
    }

    if (dates.hasMultipleDueDates()) {
      return i18n('Multiple Due Dates')
    }

    return formattedDueDateWithStatus(dates.bestDueAt(), dates.bestAvailableTo()).join('  â€¢  ')
  }

  ungradedBubble (assignment: Assignment) {
    if (!assignment.needs_grading_count || assignment.grading_type === 'not_graded') {
      return <View />
    }

    const text = i18n(`{
      count, plural,
        one {# Needs Grading}
      other {# Need Grading}
    }`, { count: assignment.needs_grading_count }).toUpperCase()
    return (
      <Text style={styles.ungradedText}>{text}</Text>
    )
  }

  render () {
    const { assignment, selected } = this.props
    return (
      <View>
        <View>
          <Row
            renderImage={this._renderIcon}
            title={assignment.name}
            titleProps={{ ellipsizeMode: 'tail', numberOfLines: 2 }}
            subtitle={this.dueDate(assignment)}
            border='bottom'
            disclosureIndicator={true}
            testID={`assignment-list.assignment-list-row.cell-${assignment.id}`}
            onPress={this.onPress}
            selected={selected}
            height='auto'
          >
            {this.ungradedBubble(assignment)}
          </Row>
        </View>
      </View>
    )
  }

  _renderIcon = () => {
    const assignment = this.props.assignment
    let image = Images.course.assignments
    let testIDSuffix = `-icon-${assignment.published ? 'published' : 'not-published'}-${assignment.id}.icon-img`
    let testID = `assignment-list-row-assignment${testIDSuffix}`
    if (assignment.submission_types.includes('online_quiz')) {
      image = Images.course.quizzes
      testID = `assignment-list-row-quiz${testIDSuffix}`
    } else if (assignment.submission_types.includes('discussion_topic')) {
      image = Images.course.discussions
      testID = `assignment-list-row-discussion${testIDSuffix}`
    }
    return (
      <View style={styles.icon} testID={testID}>
        <AccessIcon entry={assignment} tintColor={this.props.tintColor} style={styles.icon} image={image} />
      </View>
    )
  }
}

const styles = StyleSheet.create({
  ungradedText: {
    flex: 0,
    alignSelf: 'flex-start',
    fontSize: 11,
    fontWeight: '600',
    color: '#008EE2',
    borderRadius: 9,
    borderColor: '#008EE2',
    borderWidth: 1,
    backgroundColor: 'white',
    paddingTop: 2,
    paddingBottom: 2,
    paddingLeft: 6,
    paddingRight: 6,
    marginTop: 4,
    overflow: 'hidden',
  },
  icon: {
    alignSelf: 'flex-start',
  },
})
