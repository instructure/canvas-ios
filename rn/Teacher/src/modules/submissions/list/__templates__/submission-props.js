//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import template, { type Template } from '../../../../utils/template'
import type { SubmissionDataProps } from '../submission-prop-types'

const templates = {
  ...require('../../../../__templates__/submissions'),
}

const submissionWithHistory = templates.submissionHistory()

// $FlowFixMe
export const submissionProps: Template<SubmissionDataProps> = template({
  grade: 'B-',
  name: submissionWithHistory.user.name,
  status: 'submitted',
  userID: submissionWithHistory.user.id,
  avatarURL: submissionWithHistory.user.avatar_url,
  sectionID: '1',
})
