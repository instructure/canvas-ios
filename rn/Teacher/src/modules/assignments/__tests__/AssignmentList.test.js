//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/* eslint-disable flowtype/require-valid-file-annotation */

import React from 'react'
import { ActionSheetIOS, NativeModules } from 'react-native'
import { AssignmentList } from '../AssignmentList'
import { shallow } from 'enzyme'
import AssignmentListRow from '../components/AssignmentListRow'
import * as templates from '../../../__templates__/index'

jest.mock('react-native/Libraries/Components/Touchable/TouchableHighlight', () => 'TouchableHighlight')
jest.mock('react-native/Libraries/Components/Touchable/TouchableOpacity', () => 'TouchableOpacity')
jest.mock('../../../routing')

jest.mock('react-native/Libraries/ActionSheetIOS/ActionSheetIOS', () => ({
  showActionSheetWithOptions: jest.fn(),
}))

let group = templates.assignmentGroup()
let gradingPeriod = {
  ...templates.gradingPeriod(),
  assignmentRefs: [],
}
let course = templates.course()

let defaultProps = {}

beforeEach(() => {
  defaultProps = {
    courseID: course.id,
    courseName: course.name,
    assignmentGroups: [group],
    navigator: templates.navigator(),
    gradingPeriods: [gradingPeriod],
    currentGradingPeriodID: null,
    refreshAssignmentList: jest.fn(),
    updateCourseDetailsSelectedTabSelectedRow: jest.fn(),
    refresh: jest.fn(),
    refreshing: false,
    courseColor: '#fff',
    selectedRowID: templates.assignment().id,
    screenTitle: 'Assignments',
    showTotalScore: false,
    pending: 0,
    ListRow: AssignmentListRow,
    getGradesForGradingPeriod: jest.fn(),
  }
  jest.clearAllMocks()
})

describe('AssignmentList', () => {
  it('renders pending', () => {
    let tree = shallow(
      <AssignmentList {...defaultProps} pending refreshing={false} />
    )
    expect(tree.find('ActivityIndicatorView').length).toEqual(1)
  })

  it('currentGradingPeriodID is used for initial filter', () => {
    let groupOne = templates.assignmentGroup({
      id: '1',
      assignments: [ templates.assignment({ id: '1' }) ],
    })
    let groupTwo = templates.assignmentGroup({
      id: '2',
      assignments: [ templates.assignment({ id: '2' }) ],
    })
    let gradingPeriod = templates.gradingPeriod({
      assignmentRefs: ['1'],
    })

    let tree = shallow(
      <AssignmentList
        {...defaultProps}
        currentGradingPeriodID={gradingPeriod.id}
        assignmentGroups={[groupOne, groupTwo]}
        gradingPeriods={[gradingPeriod]}
      />
    )

    expect(tree.find('Heading1').props().children).toEqual(gradingPeriod.title)
    expect(tree.find('SectionList').props().sections[0].data).toEqual(groupOne.assignments)
  })

  it('currentGradingPeriodID can apply when gradingPeriods loads', () => {
    let groupOne = templates.assignmentGroup({
      id: '1',
      assignments: [ templates.assignment({ id: '1' }) ],
    })
    let groupTwo = templates.assignmentGroup({
      id: '2',
      assignments: [ templates.assignment({ id: '2' }) ],
    })
    let gradingPeriod = templates.gradingPeriod({
      assignmentRefs: ['1'],
    })

    let tree = shallow(
      <AssignmentList
        {...defaultProps}
        currentGradingPeriodID={gradingPeriod.id}
        assignmentGroups={[groupOne, groupTwo]}
        gradingPeriods={[]}
      />
    )
    tree.setProps({ gradingPeriods: [gradingPeriod] })
    expect(tree.find('Heading1').props().children).toEqual(gradingPeriod.title)
    expect(tree.find('SectionList').props().sections[0].data).toEqual(groupOne.assignments)
  })

  it('navigates to assignment', () => {
    const navigator = templates.navigator({
      show: jest.fn(),
    })
    const assignment = group.assignments[0]
    const tree = shallow(new AssignmentList({ ...defaultProps, navigator }).renderRow({ item: assignment, index: 0 }))
    let button = tree.find(`[testID="assignment-list.assignment-list-row.cell-${assignment.id}"]`)
    button.simulate('press')

    expect(navigator.show).toHaveBeenCalled()
  })

  it('navigates to graded discussion', () => {
    const navigator = templates.navigator({
      show: jest.fn(),
    })
    const assignment = Object.assign(group.assignments[0], { discussion_topic: { id: '1' } })
    const assignmentGroup = Object.assign(defaultProps.assignmentGroups[0], { assignments: [assignment] })
    const props = {
      ...defaultProps,
      assignmentGroups: [assignmentGroup],
    }
    const tree = shallow(
      new AssignmentList({ ...props, navigator }).renderRow({ item: assignment, index: 0 })
    )
    let button = tree.find(`[testID="assignment-list.assignment-list-row.cell-${assignment.id}"]`)
    button.simulate('press')
    expect(navigator.show).toHaveBeenCalledWith('/courses/987654321/discussion_topics/1')
  })

  it('navigates to quiz', () => {
    const navigator = templates.navigator({
      show: jest.fn(),
    })
    const assignment = Object.assign(group.assignments[0], { quiz_id: '1', discussion_topic: null })
    const assignmentGroup = Object.assign(defaultProps.assignmentGroups[0], { assignments: [assignment] })
    const props = {
      ...defaultProps,
      assignmentGroups: [assignmentGroup],
    }
    const tree = shallow(
      new AssignmentList({ ...props, navigator }).renderRow({ item: assignment, index: 0 })
    )
    let button = tree.find(`[testID="assignment-list.assignment-list-row.cell-${assignment.id}"]`)
    button.simulate('press')
    expect(navigator.show).toHaveBeenCalledWith(assignment.html_url)
  })

  it('hides filter button without grading periods', () => {
    let tree = shallow(
      <AssignmentList {...defaultProps} gradingPeriods={[]} />
    )
    expect(tree.find(`[testID="assignment-list.filter"]`).length).toEqual(0)
  })

  it('shows filter options', () => {
    let tree = shallow(
      <AssignmentList {...defaultProps} />
    )

    let button = tree.find('[testID="assignment-list.filter"]')
    button.simulate('press')

    expect(ActionSheetIOS.showActionSheetWithOptions).toHaveBeenCalledWith({
      options: [
        gradingPeriod.title,
        'Cancel',
      ],
      cancelButtonIndex: 1,
      title: 'Filter by:',
    }, expect.any(Function))
  })

  it('filters list by grading period', () => {
    let groupOne = templates.assignmentGroup({
      id: '1',
      assignments: [ templates.assignment({ id: '1', name: 'One' }) ],
    })
    let groupTwo = templates.assignmentGroup({
      id: '2',
      assignments: [ templates.assignment({ id: '2', name: 'Two' }) ],
    })
    let gradingPeriod = templates.gradingPeriod({
      assignmentRefs: ['1'],
    })
    let tree = shallow(
      <AssignmentList
        {...defaultProps}
        assignmentGroups={[groupOne, groupTwo]}
        gradingPeriods={[gradingPeriod]}
      />
    )
    ActionSheetIOS.showActionSheetWithOptions = jest.fn((options, fn) => fn(0))
    let filter = tree.find('[testID="assignment-list.filter"]')
    filter.simulate('press')

    let list = tree.find('SectionList')
    let sections = list.props().sections
    let row = shallow(list.props().renderItem({ item: sections[0].data[0], index: 0 })).find('Row')

    expect(tree.find('Heading1').props().children).toEqual(gradingPeriod.title)
    expect(sections.length).toEqual(1)
    expect(row.prop('title')).toEqual('One')

    ActionSheetIOS.showActionSheetWithOptions = jest.fn((options, fn) => fn(1))
    filter.simulate('press')
    expect(tree.find('SectionList').props().sections.length).toEqual(2)
  })

  it('will call refreshlist with the grading period id when it has no assignmentRefs', () => {
    let tree = shallow(
      <AssignmentList
        {...defaultProps}
      />
    )
    tree.instance().updateFilter(0)
    expect(defaultProps.refreshAssignmentList).toHaveBeenCalledWith(course.id, gradingPeriod.id, true)
  })

  it('removes filter when clear button pressed', () => {
    let tree = shallow(
      <AssignmentList {...defaultProps} />
    )

    ActionSheetIOS.showActionSheetWithOptions = jest.fn((options, fn) => fn(0))
    let filter = tree.find('[testID="assignment-list.filter"]')
    filter.simulate('press')
    let clearButton = tree.find('[testID="assignment-list.filter"]')
    expect(clearButton.props().children).toEqual('Clear filter')
    clearButton.simulate('press')
    tree.update()
    expect(tree.find('Heading1').props().children).toEqual('All')
  })

  it('renders the total grade when provided', () => {
    let tree = shallow(
      <AssignmentList {...defaultProps} currentScore={99} showTotalScore />
    )
    expect(tree.find('[testID="assignment-list.total-grade"] Text').props().children).toEqual('99%')
  })

  it('renders N/A when there is no currentScore', () => {
    let tree = shallow(
      <AssignmentList {...defaultProps} showTotalScore />
    )
    expect(tree.find('[testID="assignment-list.total-grade"] Text').props().children).toEqual('N/A')
  })

  it('shows an activity indicator when the grading period is loading', () => {
    let tree = shallow(
      <AssignmentList {...defaultProps} showTotalScore />
    )
    tree.setState({ loadingGrade: true })
    tree.update()

    expect(tree.find('ActivityIndicator').length).toEqual(1)
  })

  it('refreshes grading period grades on filter and refresh after filter', async () => {
    let grades = templates.enrollment().grades
    defaultProps.getGradesForGradingPeriod = jest.fn(() => Promise.resolve(grades))
    let tree = shallow(
      <AssignmentList {...defaultProps} showTotalScore />
    )
    ActionSheetIOS.showActionSheetWithOptions = jest.fn((options, fn) => fn(0))
    let filter = tree.find('[testID="assignment-list.filter"]')
    filter.simulate('press')
    tree.find('SectionList').simulate('Refresh')
    await tree.update()
    expect(defaultProps.getGradesForGradingPeriod).toHaveBeenCalledWith(defaultProps.courseID, 'self', gradingPeriod.id)
    expect(defaultProps.getGradesForGradingPeriod).toHaveBeenCalledTimes(2)
    expect(tree.find('[testID="assignment-list.total-grade"] Text').props().children).toEqual(`${grades.current_score}%`)
  })

  it('calls refreshAssignmentList when a filter is updated', () => {
    const tree = shallow(<AssignmentList {...defaultProps} />)
    tree.instance().updateFilter(0)
    expect(defaultProps.refreshAssignmentList).toHaveBeenCalledWith(defaultProps.courseID, gradingPeriod.id, true)
  })

  it('tries to request app store review when closing', () => {
    const tree = shallow(
      <AssignmentList
        {...defaultProps}
        currentScore={89}
        showTotalScore={false}
      />
    )
    tree.unmount()
    expect(NativeModules.AppStoreReview.handleSuccessfulSubmit)
      .not.toHaveBeenCalled()
    tree.setProps({ currentScore: 95 })
    tree.unmount()
    expect(NativeModules.AppStoreReview.handleSuccessfulSubmit)
      .not.toHaveBeenCalled()
    tree.setProps({ showTotalScore: true })
    tree.unmount()
    expect(NativeModules.AppStoreReview.handleSuccessfulSubmit)
      .toHaveBeenCalled()
  })

  it('donates siri shortcut', () => {
    defaultProps.courseCode = '123'
    defaultProps.courseID = '1'
    shallow(<AssignmentList {...defaultProps} showGrades />)
    expect(NativeModules.SiriShortcutManager.donateSiriShortcut).toHaveBeenCalledWith({
      identifier: 'com.instructure.siri.shortcut.getgrades',
      name: '123',
      url: '/courses/1/grades',
    })
  })
})
