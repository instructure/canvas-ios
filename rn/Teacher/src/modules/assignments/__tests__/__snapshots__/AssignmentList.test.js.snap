// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Screen
  navBarColor="#fff"
  navBarStyle="dark"
  onTraitCollectionChange={[Function]}
  showDismissButton={true}
  statusBarStyle="default"
  subtitle="Learn React Native"
  testID="assignment-list"
  title="Assignments"
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
    testID="assignment-list.container-view"
  >
    <View
      style={
        Object {
          "borderBottomColor": "#C7CDD1",
          "borderBottomWidth": 0.5,
          "paddingBottom": 8,
          "paddingHorizontal": 16,
          "paddingTop": 16,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-end",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
        testID="assignment-list.filter-header-view"
      >
        <Heading1
          style={
            Object {
              "flex": 1,
            }
          }
          testID="assignment-list.filter-title-lbl"
        >
          All Grading Periods
        </Heading1>
        <LinkButton
          onPress={[Function]}
          style={
            Object {
              "marginBottom": 1,
            }
          }
          testID="assignment-list.filter"
        >
          Filter
        </LinkButton>
      </View>
    </View>
    <SectionList
      ListEmptyComponent={
        <ListEmptyComponent
          title="There are no assignments to display."
        />
      }
      keyExtractor={[Function]}
      onRefresh={[Function]}
      refreshing={false}
      renderItem={[Function]}
      renderSectionHeader={[Function]}
      sections={
        Array [
          Object {
            "assignments": Array [
              Object {
                "anonymize_students": false,
                "assignment_group_id": "1",
                "automatic_peer_reviews": false,
                "course_id": "987654321",
                "created_at": "2017-03-17T19:15:25Z",
                "description": "a course that can take any beginner to professional in a matter of microseconds.  Harness the power of your favorite languages like PHP and Cold Fusion.",
                "due_at": "2037-06-01T05:59:00Z",
                "grade_group_students_individually": false,
                "grading_type": "points",
                "group_category_id": null,
                "has_overrides": false,
                "html_url": "http://hostname.tld/whatever",
                "id": "123456789",
                "lock_at": "2037-06-01T05:59:59Z",
                "name": "learn to write code",
                "needs_grading_count": 0,
                "only_visible_to_overrides": false,
                "overrides": Array [],
                "peer_reviews": false,
                "points_possible": 20,
                "position": 1,
                "published": true,
                "rubric": null,
                "rubric_assessment": Object {},
                "rubric_settings": null,
                "submission_types": Array [
                  "online_text_entry",
                  "online_url",
                  "media_recording",
                  "online_upload",
                ],
                "unlock_at": "2017-03-17T06:00:00Z",
                "unpublishable": false,
                "updated_at": "2017-03-17T19:15:25Z",
              },
            ],
            "data": Array [
              Object {
                "anonymize_students": false,
                "assignment_group_id": "1",
                "automatic_peer_reviews": false,
                "course_id": "987654321",
                "created_at": "2017-03-17T19:15:25Z",
                "description": "a course that can take any beginner to professional in a matter of microseconds.  Harness the power of your favorite languages like PHP and Cold Fusion.",
                "due_at": "2037-06-01T05:59:00Z",
                "grade_group_students_individually": false,
                "grading_type": "points",
                "group_category_id": null,
                "has_overrides": false,
                "html_url": "http://hostname.tld/whatever",
                "id": "123456789",
                "lock_at": "2037-06-01T05:59:59Z",
                "name": "learn to write code",
                "needs_grading_count": 0,
                "only_visible_to_overrides": false,
                "overrides": Array [],
                "peer_reviews": false,
                "points_possible": 20,
                "position": 1,
                "published": true,
                "rubric": null,
                "rubric_assessment": Object {},
                "rubric_settings": null,
                "submission_types": Array [
                  "online_text_entry",
                  "online_url",
                  "media_recording",
                  "online_upload",
                ],
                "unlock_at": "2017-03-17T06:00:00Z",
                "unpublishable": false,
                "updated_at": "2017-03-17T19:15:25Z",
              },
            ],
            "group_weight": 1,
            "id": "1",
            "integration_data": Object {},
            "key": "1",
            "name": "Learn React Native",
            "position": 1,
            "sis_source_id": "",
          },
        ]
      }
      testID="assignment-list.list"
    />
  </View>
</Screen>
`;
